splice_machine/src/main/java/com/splicemachine/derby/impl/sql/catalog/SpliceDataDictionary.java

```diff
@@ -1715,46 +1719,45 @@ public void addMinRetentionPeriodColumn(TransactionController tc) throws Standar
+        FormatableBitSet columnToReadSet = new FormatableBitSet(SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT);
+        FormatableBitSet columnToUpdateSet = new FormatableBitSet(SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT);
+        for (int i = 0; i < SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT; i++) {
+            // partial row updates do not work properly (DB-9388), therefore, we read all columns and mark them all for
+            // update even if this is not necessary for all of them.
+            columnToReadSet.set(i);
             columnToUpdateSet.set(i);
         }
-        /* Set up a couple of row templates for fetching CHARS */
+        /* Set up a row template for fetching */
         DataValueDescriptor[] rowTemplate = new DataValueDescriptor[SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT];
-        DataValueDescriptor[] replaceRow= new DataValueDescriptor[SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT];
+        /* Set up another row for replacing the existing row, effectively updating it */
+        DataValueDescriptor[] replaceRow = new DataValueDescriptor[SYSALIASESRowFactory.SYSALIASES_COLUMN_COUNT];
 
         /* Scan the entire heap */
-        ScanController sc=
-                tc.openScan(
-                        ti.getHeapConglomerate(),
-                        false,
-                        TransactionController.OPENMODE_FORUPDATE,
-                        TransactionController.MODE_TABLE,
-                        TransactionController.ISOLATION_REPEATABLE_READ,
-                        columnToReadSet,
-                        null,
-                        ScanController.NA,
-                        null,
-                        null,
-                        ScanController.NA);
-
-        while(sc.fetchNext(rowTemplate)){
-            /* If JAVACLASSNAME was set to null, rewrite it to "NULL" string literal instead. */
-            for (int i=0; i<rowTemplate.length; i++) {
-                if (i+1 == SYSALIASESRowFactory.SYSALIASES_JAVACLASSNAME)
-                    if(replaceRow[i] == null) {
-                        replaceRow[i] = new SQLLongvarchar("NULL");
-                    }
+        ScanController sc = tc.openScan(
+                ti.getHeapConglomerate(),
+                false,
+                TransactionController.OPENMODE_FORUPDATE,
+                TransactionController.MODE_TABLE,
+                TransactionController.ISOLATION_REPEATABLE_READ,
+                columnToReadSet,
+                null,
+                ScanController.NA,
+                null,
+                null,
+                ScanController.NA);
+
+        while (sc.fetchNext(rowTemplate)) {
+            for (int i = 0; i < rowTemplate.length; i++) {
+                replaceRow[i] = rowTemplate[i].cloneValue(false);
+                /* If JAVACLASSNAME was set to null, rewrite it to "NULL" string literal instead. */
+                if (i + 1 == SYSALIASESRowFactory.SYSALIASES_JAVACLASSNAME && rowTemplate[i].isNull()) {
```

Comment:
Shouldn't we use the check "rowTemplate[I] == null"? I think we only have problem when this field is not populated at all, and thus null. Do we have a problem if rowTemplate[I] has a NULL value?

Reply:
The problem we’re facing here is that we store null in a non-nullable column. The column type is defined as the following:
`SystemColumnImpl.getColumn("JAVACLASSNAME", java.sql.Types.LONGVARCHAR, false /*isNullable?*/, Integer.MAX_VALUE)`
Later on, we create a DVD and store a null into it:
`row.setColumn(SYSALIASES_JAVACLASSNAME, dvf.getLongvarcharDataValue(javaClassName)); // javaClassName is null`
I think this is causing the decoder to be confused leading to corrupted decoded data since we send `null` value in a non-nullable field.
To solve this problem, I made sure that we’re writing `"NULL"` literal leaving no chance for the field to be `null`.
I think the interesting part is we can provide a much simpler solution by simply making the column nullable! i.e. it is a one-liner change from:
`SystemColumnImpl.getColumn("JAVACLASSNAME", java.sql.Types.LONGVARCHAR, false /*isNullable?*/, Integer.MAX_VALUE)`
to
`SystemColumnImpl.getColumn("JAVACLASSNAME", java.sql.Types.LONGVARCHAR, true /*isNullable?*/, Integer.MAX_VALUE)`
This makes the decoder happy and everything works as expected (just tested it). Moreover, I don’t think we need an upgrade script for it.
However it means that we will have to change the `SYSALIASES` schema. Not sure if this is something we want to do, that’s why I lean towards the proposed solution in this PR.

Reply:
Thank you @hatyo for the explanation! I'm good with the current approach taken and will approve the PR.